File: irrgarten_ruby/monster.rb
# frozen_string_literal: true

module Irrgarten
  require_relative 'dice'
  class Monster
    INITIAL_HEALTH = 5

    def initialize(name, intelligence, strength)
      @name = name
      @intelligence = intelligence
      @health = INITIAL_HEALTH
      @strength = strength
      @row = -1
      @col = -1
    end

    def dead
      @health.zero?
    end

    def attack
      Dice.intensity(@strength)
    end

    def defend(received_attack)
      is_dead = dead

      if is_dead == false
        defensive_energy = Dice.intensity(@intelligence)

        if defensive_energy < received_attack
          got_wounded
          is_dead = dead
        end
      end

      is_dead
    end

    def set_pos(row, col)
      @row = row
      @col = col
    end

    def to_s
      header_string = "#{@name}\n"
      location_string = "POSITION: [#{@row}, #{@col}]\n"
      strength_string = "STRENGTH: #{@strength}\n"
      health_string = "HEALTH: #{@health}\n"
      intelligence_string = "INTELLIGENCE: #{@intelligence}\n\n"

      header_string + location_string + strength_string + health_string + intelligence_string
    end

    def got_wounded
      @health -= 1
    end
  end
end

File: irrgarten/src/irrgarten/Monster.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

/**
 *
 * @author marco
 */
public class Monster {
    private static final int INITIAL_HEALTH = 5;
    private final String name;
    private final float intelligence;
    private final float strength;
    private float health;
    private int row;
    private int col;
    
    public Monster(String name, float intelligence, float strength){
        this.name = name;
        this.intelligence = intelligence;
        this.health = INITIAL_HEALTH;
        this.strength = strength;
        this.row = -1;
        this.col = -1;
    }
    
    public boolean dead(){
        return health == 0;
    }
    
    public float attack(){
        return Dice.intensity(strength);
    }
    
    public boolean defend(float receivedAttack){
        boolean isDead = this.dead();
        
        if (!isDead){
            float defensiveEnergy = Dice.intensity(intelligence);
            
            if (defensiveEnergy < receivedAttack){
                gotWounded();
                isDead = this.dead();
            } 
        }
        
        return isDead;
    }
    
    public void setPos(int row, int col){
        this.row = row;
        this.col = col;
    }
    
    
    public String toString(){
        String header = "MONSTER " + this.name + "\n";
        String location = "POSTION: [ "+ this.row + ", " + this.col + " ]\n";
        String strengthString = "STRENGTH: " + this.strength + "\n";
        String healthString = "HEALTH: " + this.health + "\n";
        String intelligenceString = "INTELLIGENCE: " + this.intelligence + "\n";
        
        return header + location + strengthString + healthString + intelligenceString;
    }
    
    private void gotWounded(){
        health --;
    }
}
